### PR TITLE
#40  ExtendExceptionHandler

### PR DESCRIPTION
**Pull request by deivit98**

### CODE REVIEW

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/BusinessLogicException.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Text.Json;
+
+namespace SportsHub.Api.Exceptions.CustomExceptionModels;
+
+public class BusinessLogicException: CustomException
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/BusinessLogicException.cs:5]: Exception related to business logic shouldn't be on API layer.

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/BusinessLogicException.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public BusinessLogicException(string message)
+    {
+        Message = message;
+    }
+
+    public int StatusCode { get; } = 400;
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/BusinessLogicException.cs:12]: It is not the responsibility of the exception to set the status code.

File: SportsHub/.idea/.idea.SportsHub/.idea/encodings.xml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
```
Comment (NataliiaBidiuk) [inline SportsHub/.idea/.idea.SportsHub/.idea/encodings.xml:2]: check, please, why are these files here

File: SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs [hunk 4]
```diff
@@ -1,20 +1,55 @@
...
     public async Task InvokeAsync(HttpContext context, RequestDelegate next)
     {
         try
         {
             await next(context);
         }
-
-        catch (Exception e)
+        catch (BusinessLogicException ex)
+        {
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs:16]: Do you need these 2 catch blocks, or is it possible to combine them?

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/CustomException.cs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+﻿namespace SportsHub.Api.Exceptions.CustomExceptionModels;
+
+public abstract class CustomException: Exception
+{
+    public int StatusCode { get; set; }
+    public string Message { get; set; }
+    public abstract override string ToString();
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/CustomException.cs:7]: why do you need this to be abstract if there are the same implementations everywhere

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/CustomException.cs [hunk 6]
```diff
@@ -0,0 +1,8 @@
+﻿namespace SportsHub.Api.Exceptions.CustomExceptionModels;
+
+public abstract class CustomException: Exception
+{
+    public int StatusCode { get; set; }
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/CustomException.cs:5]: I think here better to keep an ErrorCode, not status

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/NotFoundException.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+namespace SportsHub.Api.Exceptions.CustomExceptionModels;
+
+public class NotFoundException: CustomException
+{
+    public NotFoundException(string message)
+    {
+        Message = message;
+    }
+
+
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/NotFoundException.cs:12]: extra lines

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/CustomException.cs [hunk 8]
```diff
@@ -0,0 +1,8 @@
+﻿namespace SportsHub.Api.Exceptions.CustomExceptionModels;
+
+public abstract class CustomException: Exception
+{
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/CustomException.cs:4]: I am not sure that this class should be an abstract

File: SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs (hunk 4 above)
Comment (deivit98) [reply SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs:16]: Yes i forgot to change it there . With my abstract class i can catch with CustomException and then in my private method to set which one should be .

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/NotFoundException.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+    public int StatusCode { get; } = 404;
+    public string Message { get; set; }
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/NotFoundException.cs:15]: you have these properties in the base class already

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/BusinessLogicException.cs (hunk 2 above)
Comment (deivit98) [reply SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/BusinessLogicException.cs:12]: I will set it in constructor

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/BusinessLogicException.cs (hunk 1 above)
Comment (deivit98) [reply SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/BusinessLogicException.cs:5]: Can u point me where to move it and also where i need to throw exceptions . I was thinking to throw them in repositories

File: SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs [hunk 10]
```diff
@@ -1,20 +1,55 @@
...
+    private async Task HandleExceptionAsync(HttpContext context, Exception ex)
+    {
+        string result = null;
+        context.Response.ContentType = "application/json";
+        
+        switch (ex)
+        {
+            case BusinessLogicException:
+            {
+                _exception = new BusinessLogicException("There has been error in you business logic");
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs:39]: is this switch for the test purpose only?
in the InvokeAsync you call this method for a different exception. and here again, you have to check for the exception type. Make this method generic, please.

File: SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs (hunk 4 above)
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs:16]: you can use "When" here

**Review by NataliiaBidiuk — CHANGES_REQUESTED** [review]

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/BusinessLogicException.cs (hunk 1 above)
Comment (NataliiaBidiuk) [reply SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/BusinessLogicException.cs:5]: A business logic exception should be thrown on the business layer - here is a Service layer.
Exception class should be placed on the Domain level

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/CustomException.cs (hunk 6 above)
Comment (NataliiaBidiuk) [reply SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/CustomException.cs:5]: Add the possibility for the business logic exception, for example, to be more generic. Allow setting not only a message but an error code as well

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/BusinessLogicException.cs (hunk 2 above)
Comment (MBocharskyi) [reply SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/BusinessLogicException.cs:12]: I'm talking about a different thing. An exception should have information about the error, not the status code to return. You can set error code in exception middleware.

File: SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs (hunk 10 above)
Comment (deivit98) [reply SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs:39]: Yea it was for testing purpose only . The idea of was seeing some basic concept of what my idea is . I will make it Generic or i can make it depend on abstraction

File: SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs (hunk 4 above)
Comment (deivit98) [reply SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs:16]: I totaly forgot about this keyword . I will improve it

File: SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/CustomException.cs (hunk 6 above)
Comment (deivit98) [reply SportsHub/SportsHub.Api/Exceptions/CustomExceptionModels/CustomException.cs:5]: Okey i will create additional constructor for this

File: SportsHub/.idea/.idea.SportsHub/.idea/encodings.xml (hunk 3 above)
Comment (deivit98) [reply SportsHub/.idea/.idea.SportsHub/.idea/encodings.xml:2]: I staged them by mistake when i pushed from my Rider . But they should probably be ignored from what i have seen they are in the gitignore

File: SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs [hunk 11]
```diff
@@ -10,11 +11,34 @@ public async Task InvokeAsync(HttpContext context, RequestDelegate next)
...
+    private async Task HandleNotFoundExceptionAsync(HttpContext context, NotFoundException ex)
+    {
+        context.Response.ContentType = "/application/json";
+        context.Response.StatusCode = ex.ErrorCode;
+        string result = ex.ToString();
+        await context.Response.WriteAsync(result);
+    }
+
+    private async Task HandleBusinessLogicExceptionAsync(HttpContext context, BusinessLogicException ex)
+    {
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs:38]: what is different between these methods? HandleNotFoundExceptionAsync and current?

File: SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs [hunk 12]
```diff
@@ -10,11 +11,34 @@ public async Task InvokeAsync(HttpContext context, RequestDelegate next)
...
             context.Response.StatusCode = (int)HttpStatusCode.InternalServerError;
-            await context.Response.WriteAsync(e.Message);
+            await context.Response.WriteAsync(ex.Message);
         }
     }
+
+    private async Task HandleNotFoundExceptionAsync(HttpContext context, NotFoundException ex)
+    {
+        context.Response.ContentType = "/application/json";
+        context.Response.StatusCode = ex.ErrorCode;
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs:32]: in the Response.StatusCode = should be HTTPStatus code, 400,404, 500
Error code should be returned with the message in response bode

File: SportsHub/SportsHub.Core/Models/CustomExceptionModels/BusinessLogicException.cs [hunk 13]
```diff
@@ -0,0 +1,23 @@
+﻿namespace SportsHub.Api.Exceptions.CustomExceptionModels;
+
+public class BusinessLogicException: Exception
+{
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Core/Models/CustomExceptionModels/BusinessLogicException.cs:4]: why have you decided to get rid of CoreException?

File: SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs (hunk 12 above)
Comment (deivit98) [reply SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs:32]: Oh okey i understand and i will change it .

File: SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs (hunk 11 above)
Comment (deivit98) [reply SportsHub/SportsHub.Api/Exceptions/ExceptionHandler.cs:38]: I probably forgot to change it . Because before i tried to throw exceptions in business layer i didnt had the ExceptionMessages class and i was setting different messages in this methods for testing . After i got home from work i will see to it

File: SportsHub/SportsHub.Core/Models/CustomExceptionModels/BusinessLogicException.cs (hunk 13 above)
Comment (deivit98) [reply SportsHub/SportsHub.Core/Models/CustomExceptionModels/BusinessLogicException.cs:4]: Maybe I misunderstood you , but I thought you didnt wanted Abstract class

**Review by MBocharskyi — APPROVED** [review]

File: SportsHub/SportsHub.Core/Models/CustomExceptionModels/BusinessLogicException.cs (hunk 13 above)
Comment (deivit98) [reply SportsHub/SportsHub.Core/Models/CustomExceptionModels/BusinessLogicException.cs:4]: Okey i will add abstract class

Comment (deivit98) [reply SportsHub/SportsHub.Core/Models/CustomExceptionModels/BusinessLogicException.cs:4]: So i need to implement a CoreException which inherits the Exception then our CustomExceptions will inherit the Core , but it shoudnt be a abstract class ? Did i get it right

File: SportsHub/SportsHub.Application/Services/ArticleService.cs [hunk 14]
```diff
@@ -14,7 +16,8 @@ public ArticleService(IUnitOfWork unitOfWork)
 
         public async Task<Article?> GetByTitleAsync(string title)
         {
-            return await _unitOfWork.ArticleRepository.GetByTitleAsync(title);
+            return await _unitOfWork.ArticleRepository.GetByTitleAsync(title) ??
+                   throw new NotFoundException(404, ExceptionMessages.NotFound);
```
[the remark this answers is not in the document]
Comment (deivit98) [reply SportsHub/SportsHub.Application/Services/ArticleService.cs:20]: Okey

File: SportsHub/.idea/.idea.SportsHub/.idea/encodings.xml (hunk 3 above)
Comment (NataliiaBidiuk) [reply SportsHub/.idea/.idea.SportsHub/.idea/encodings.xml:2]: but they also still here

File: SportsHub/SportsHub.Application/Services/ArticleService.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (deivit98) [reply SportsHub/SportsHub.Application/Services/ArticleService.cs]: implemented it

File: SportsHub/SportsHub.Core/Models/CustomExceptionModels/BusinessLogicException.cs (hunk 13 above)
Comment (NataliiaBidiuk) [reply SportsHub/SportsHub.Core/Models/CustomExceptionModels/BusinessLogicException.cs:4]: yes, so you do not need to duplicate properties

Comment (deivit98) [reply SportsHub/SportsHub.Core/Models/CustomExceptionModels/BusinessLogicException.cs:4]: okey i will .

**Review by NataliiaBidiuk — APPROVED** [review]

**Comment by MBocharskyi** [comment]
Please resolve conflicts.
